ShareExtension/ShareExtension/Views/ShareView.swift

```diff
@@ -16,22 +17,52 @@ struct ShareView: View {
 
     var body: some View {
         VStack {
-            if let image = viewModel.image {
-                Image(uiImage: image)
-                    .resizable()
-                    .clipShape(Rectangle())
-                    .aspectRatio(contentMode: .fit)
-                    .frame(width: Numbers.sharedImageSize, height: Numbers.sharedImageSize)
-            }
-            Button {
-                viewModel.saveDataToDocuments(viewModel.image?.pngData())
-            } label: {
-                Text(Strings.shareMessage)
-                    .bold()
-                    .foregroundStyle(.black)
-                    .padding()
-                    .background(.yellow)
-            }
+            imagePreview()
+            saveButton
+        }
+    }
+}
+
+// MARK: - Private Views
+
+private extension ShareView {
+
+    @ViewBuilder func imagePreview() -> some View {
+        if let image = viewModel.image {
+            Image(uiImage: image)
+                .resizable()
+                .clipShape(Rectangle())
+                .aspectRatio(contentMode: .fit)
+                .frame(width: Numbers.sharedImageSize, height: Numbers.sharedImageSize)
+        }
+    }
+
+    var saveButton: some View {
+        Button {
+            didPressSave()
+        } label: {
+            Text(Strings.shareMessage)
+                .bold()
+                .foregroundStyle(.black)
+                .padding()
+                .background(showProgressView ? .gray : .yellow)
+        }
+        .overlay {
+            ProgressView()
+                .opacity(showProgressView ? 1 : 0)
+        }
+        .disabled(showProgressView)
+    }
+}
+
+// MARK: - Private Methods
+
+private extension ShareView {
+    func didPressSave() {
+        showProgressView = true
+        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
```

Comment:
Maybe as an idea to create random number between 1-3 seconds for adding extra wait to somehow simulate different wait times but also is good like it is :)